Fabric.Authorization.AccessControl/src/app/modules/access-control/member-add/member-add.component.ts

```diff
@@ -1,129 +1,162 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, OnDestroy } from '@angular/core';
+
+import { Subject } from 'rxjs/Subject';
+import { Subscription } from 'rxjs/Subscription';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/takeUntil';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/observable/throw';
+
 import {
   FabricExternalIdpSearchService,
   FabricAuthRoleService,
   AccessControlConfigService,
   FabricAuthUserService,
   FabricAuthGroupService
 } from '../../../services';
-import {
-  FabricPrincipal,
-  Role,
-  User,
-  Group,
-  IdPSearchRequest
-} from '../../../models';
+import { IFabricPrincipal, IRole, IUser, IGroup } from '../../../models';
+import { inherits } from 'util';
+import { Router } from '@angular/router';
+
+interface IRoleModel extends IRole {
+  selected: boolean;
+}
+
+interface IFabricPrincipalModel extends IFabricPrincipal {
+  selected: boolean;
+}
 
 @Component({
   selector: 'app-member-add',
   templateUrl: './member-add.component.html',
-  styleUrls: ['./member-add.component.css']
+  styleUrls: ['./member-add.component.scss']
 })
-export class MemberAddComponent implements OnInit {
-  public searchInput: string;
-  public principals: Array<FabricPrincipal>;
-  public selectedPrincipal: FabricPrincipal;
-  public roles: Array<Role>;
-  public selectedRoles: Array<Role>;
+export class MemberAddComponent implements OnInit, OnDestroy {
+  public principals: Array<IFabricPrincipalModel>;
+  public roles: Array<IRoleModel>;
+  public searching = false;
+
+  public searchText = new Subject<string>();
+  private ngUnsubscribe: any = new Subject();
+  private principalSelected = false;
 
   constructor(
     private idpSearchService: FabricExternalIdpSearchService,
     private roleService: FabricAuthRoleService,
     private userService: FabricAuthUserService,
     private configService: AccessControlConfigService,
-    private groupService: FabricAuthGroupService
+    private groupService: FabricAuthGroupService,
+    private router: Router
   ) {
-    this.selectedRoles = [];
   }
 
   ngOnInit() {
-    this.getRoles();
-  }
+    // Roles
+    this.roleService
+      .getRolesBySecurableItemAndGrain(
+        this.configService.grain,
+        this.configService.securableItem
+      )
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe(roleResults => {
+        this.roles = roleResults.map(role => {
+          // TODO: find user and associated roles on server
+          const newRole = <IRoleModel>role;
+          newRole.selected = false;
+          return newRole;
+        });
+      });
 
-  onKey(searchText) {
-    if (searchText.length < 2) {
-      return;
-    }
-    const request = new IdPSearchRequest(searchText);
+    // Search text
+    this.searchText
+      .takeUntil(this.ngUnsubscribe)
```

Comment:
I didn't know about that.  Cool!

Reply:
Peter said that this was the recommended way Ward Bell said to unsubscribe from subscriptions. At least last year....